### PR TITLE
[KEYCLOAK-4374] Support SAML 2.0 AttributeValue of type anyType and nil in Assertions

### PR DESCRIPTION
**Pull request by pskopek**
https://github.com/keycloak/keycloak/pull/4368

Fixing issue when SAML20 attribute "AttributeValue" is of type "anyType". On the way I have added handling of nil attribute as well.
Commit contains tests to check the handling.

### CODE REVIEW

File: saml-core/src/test/java/org/keycloak/saml/processing/core/parsers/saml/SAMLParserTest.java [hunk 1]
```diff
@@ -216,4 +223,102 @@ public void testEmptyAttributeValueLast() throws Exception {
...
+            for (AttributeStatementType.ASTChoiceType choiceType: attributeStatementType.getAttributes()) {
+                AttributeType attr = choiceType.getAttribute();
+                String attrName = attr.getName();
+                String attrValueStatement = "unexpected value of attribute " + attrName + " of " + fileName;
+                String attrTypeStatement = "unexpected type of attribute " + attrName + " of " + fileName;
+                // test selected attributes
+                if (attrName.equals("portal_id")) {
+                    assertTrue(attrValueStatement, attr.getAttributeValue().get(0).equals("060D00000000SHZ"));
+                } else if (attrName.equals("organization_id")) {
+                    assertTrue(attrTypeStatement, attr.getAttributeValue().get(0) instanceof String);
```
Comment (hmlnarik) [inline saml-core/src/test/java/org/keycloak/saml/processing/core/parsers/saml/SAMLParserTest.java]: `Assert.assertThat(attrTypeStatement, attr.getAttributeValue().get(0), Matchers.isA(String))` gives better info on failure.

File: saml-core/src/test/java/org/keycloak/saml/processing/core/parsers/saml/SAMLParserTest.java [hunk 2]
```diff
@@ -216,4 +223,102 @@ public void testEmptyAttributeValueLast() throws Exception {
...
+                    String value = (String)attr.getAttributeValue().get(0);
+                    assertTrue(attrValueStatement, value.contains("<n3:stuff xmlns:n3=\"ftp://example.org\">00DD0000000F7L5</n3:stuff>"));
```
Comment (hmlnarik) [inline saml-core/src/test/java/org/keycloak/saml/processing/core/parsers/saml/SAMLParserTest.java]: Similarly, to get better output, use `Assert.assertThat(attrValueStatement, value, containsString("..."));`

File: saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/util/SAMLParserUtil.java [hunk 3]
```diff
@@ -316,25 +334,49 @@ public static Object parseAttributeValue(XMLEventReader xmlEventReader) throws P
...
+                event = xmlEventReader.nextTag();
+                StringWriter sw = new StringWriter();
+                XMLEventWriter writer = XMLOutputFactory.newInstance().createXMLEventWriter(sw);
+                QName tagName = event.asStartElement().getName();
+                boolean endTagReached = false;
+                do {
+                    writer.add(event);
+                    event = (XMLEvent) xmlEventReader.next();
+                    endTagReached = event.isEndElement() && event.asEndElement().getName().equals(tagName);
+                } while (xmlEventReader.hasNext() && !endTagReached);
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/util/SAMLParserUtil.java]: Would the following xml snippet be processed correctly?
```xml
<status>
  <code>
    <status>XYZ</status>
  <code>
</status>
```
Please add a testcase for that.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Please see inline comments.

**Comment by pskopek** [comment]
@hmlnarik all your concerns fixed. Can you review again, please?

**Review by hmlnarik — APPROVED** [review]
Thanks, looks good to me.
@stianst This PR can be merged.

**Pull request merged**